<h5 class="text-danger">Search for Gene by Primary Annotation</h5>
<form [formGroup]="queryForm" (ngSubmit)="processQuery()">
    <div class="form-group row">
        <label for="organism_name" class="col-md-3 col-form-label">Select an organism</label>
        <div class="col-md-6">
            <select formControlName="organism_name" id="organism_name" class="custom-select">
                <option value=""></option>
                <option *ngFor="let item of listItems" value="{{ item.organism + '-' + item.version }}">{{ item.label }}</option>
            </select>
        </div>
    </div>
    <div class="form-group row">
        <label for="primary_annotation" class="col-md-3 col-form-label">Primary Annotation Name</label>
        <div class="col-md-6">
            <input type="text" class="form-control" formControlName="primary_annotation" id="primary_annotation">
        </div>
    </div>
    <button type="submit" class="btn btn-primary px-3" [disabled]="!queryForm.valid || processingQuery"><span *ngIf="!processingQuery"><i class="fas fa-search"></i> Search</span><span *ngIf="processingQuery">Please Wait...</span></button>
</form>

<ng-container *ngIf="queryItem.length">
    <hr>
    <h3 class="text-primary">Search Results <span *ngIf="queryItem">({{ queryItem.length }})</span></h3>
    <table id="primary-annotation-search-result" class="table table-striped table-sm my-3"></table>
</ng-container>

<div class="alert alert-danger mt-3" role="alert">primary annotations are derived from the blast hits of the protein sequences with nr database having > 50% identity over 50% of the query length. If the first hit matches with a un-named protein product, then the annotation of the subsequent qualified hit is assigned as the primary annotation for the query sequence.<br>Input Example: Transcript keyword: actin</div>
